core/src/smartcontracts/wasm.rs

```diff
@@ -465,55 +547,73 @@ impl<'wrld> Runtime<'wrld> {
         bytes: impl AsRef<[u8]>,
         state: State,
     ) -> Result<(), Error> {
-        let mut store = self.create_store(state).map_err(Error::Instantiation)?;
-
-        let smart_contract = self
-            .create_smart_contract(&mut store, bytes)
-            .map_err(Error::Instantiation)?;
-
-        let account_bytes = account_id.encode();
-        let account_bytes_len = account_bytes
-            .len()
-            .try_into()
-            .wrap_err(format!(
-                "Encoded account ID has size larger than {}::MAX",
-                std::any::type_name::<WasmUsize>()
-            ))
-            .map_err(Error::Other)?;
+        let mut store = self.create_store(state)?;
+        let smart_contract = self.create_smart_contract(&mut store, bytes)?;
 
-        let account_offset = {
+        let (account_offset, account_bytes_len) = {
             let alloc_fn = smart_contract
-                .get_typed_func::<WasmUsize, WasmUsize, _>(&mut store, WASM_ALLOC_FN)
+                .get_typed_func(&mut store, exported::WASM_ALLOC_FN)
                 .map_err(Error::ExportNotFound)?;
 
-            let acc_offset = alloc_fn
-                .call(&mut store, account_bytes_len)
-                .map_err(Error::ExportFnCall)?;
-
-            smart_contract
-                .get_memory(&mut store, WASM_MEMORY_NAME)
-                .ok_or_else(|| {
-                    Error::ExportNotFound(anyhow::Error::msg(format!(
-                        "{}: export not found or not a memory",
-                        WASM_MEMORY_NAME
-                    )))
-                })?
-                .write(&mut store, acc_offset as usize, &account_bytes)
-                .map_err(|error| Trap::new(error.to_string()))?;
+            let memory = Self::get_smartcontract_memory(&smart_contract, &mut store)?;
 
-            acc_offset
+            Self::encode_to_memory(account_id, &memory, &alloc_fn, &mut store)?
```

Comment:
```suggestion
            Self::encode_to_memory(account_id, &alloc_fn, &mut store)?
```
also seems to me that you can get memory from inside the function

Reply:
although it's a little bit weird that you get memory from caller when decoding and from store when encoding. As if something's not quite right with abstraction. Idk, if it can be made better

Reply:
This is a code smell, but I doubt that it can leak memory. 

New ticket?

Reply:
> also seems to me that you can get memory from inside the function

See [this comment](https://github.com/hyperledger/iroha/pull/2569#discussion_r936034813).
Also as an example of writing two times see `execute_trigger()`

Reply:
> although it's a little bit weird that you get memory from caller when decoding and from store when encoding. As if something's not quite right with abstraction. Idk, if it can be made better

Hm... All I've done is just refactoring by looking and the previous existed code and trying to avoid repeating it by encapsulating in functions when I was writing the `execute_trigger()` function. I'm not really sure what does `store` means here. If you have some ideas, as Aleksandr suggested, let's create a new issue. I'm just have no idea how to describe this issue

Reply:
I think it's ok

Reply:
I think this will go away if instead of pushing the event into triggers unconditionally, we instead wrap syntactic sugar around querying.